frontend/src/hooks/usePaginationRange.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * Custom hook to generate page ranges for any pagination component.
+ * Referenced from
+ * https://www.freecodecamp.org/news/build-a-custom-pagination-component-in-react/
+ */
+
+import { useMemo } from 'react'
+import range from 'lodash/range'
+
+interface UsePaginationRangeProps<T extends unknown = string> {
+  /**
+   * Number of pages to display to left and right of current page.
+   */
+  siblingCount: number
+
+  /**
+   * Total number of elements to paginate.
+   */
+  totalCount: number
+
+  /**
+   * Size of each page. Determines the number of rendered page counts.
+   */
+  pageSize: number
+
+  /**
+   * Represents the current active page. Note that a `1-based index` is used.
+   */
+  currentPage: number
+
+  /**
+   * A constant to denote a separator in the pagination component.
+   */
+  separator: T
+}
+
+export const usePaginationRange = <T extends unknown = string>({
+  totalCount,
+  pageSize,
+  siblingCount,
+  currentPage,
+  separator,
+}: UsePaginationRangeProps<T>): (T | number)[] => {
+  const paginationRange = useMemo(() => {
+    const totalPageCount = Math.ceil(totalCount / pageSize)
+
+    // Pages count is determined as siblingCount + firstPage + lastPage + currentPage + 2*separator
+    const totalPageNumbers = siblingCount + 5
+
+    /**
+     * Case 1:
+     * If the number of pages is less than the page numbers we want to show in
+     * the component, we return the range [1..totalPageCount]
+     */
+    if (totalPageNumbers >= totalPageCount) {
+      return range(1, totalPageCount + 1)
+    }
+
+    // Calculate left and right sibling index and make sure they are within
+    // range [1..totalPageCount]
+    const leftSiblingIndex = Math.max(currentPage - siblingCount, 1)
+    const rightSiblingIndex = Math.min(
+      currentPage + siblingCount,
+      totalPageCount,
+    )
+
+    // Do not show separator just when there is just one page number to be
+    // inserted between the extremes of sibling and the page limits
+    // i.e 1 and totalPageCount.
+    const shouldShowLeftSeparator = leftSiblingIndex - 1 > 2
```

Comment:
I think `leftSiblingIndex > 3` would be clearer here, so it's obvious that we only show the left separator if the left sibling is at least 4

Reply:
actually, rather than constantly using 3 as a magic number, it might be good to declare it as a constant or even a prop. something like `MIN_ITEMS_WITHOUT_SEP` or `minItemsWithoutSep` if it's a prop, since it refers to the minimum number of items on a side which doesn't have a separator.

Reply:
I'm thinking 3 is not really the value the code cares about, rather the logic the code tries to convey is that a separator represents a gap of 2 or above.

I imagine that's why the formula is showing the `> 2` above, but then it is unfortunate the formula for the right separator below  is not following the same formula format.

I reckon it would be clearer if both formulas were consistent, like this:
```javascript
const shouldShowLeftSeparator = leftSiblingIndex - 1 > 2
const shouldShowRightSeparator = totalPageCount - rightSiblingIndex > 2
```

And then we can replace the `2` with a named variable, perhaps named `minGapSize`, or such like.

Reply:
correct; we want to only show if the differences between the ends are larger than a gap of 2; will create a named variable.